ldap/servers/plugins/replication/repl5_agmt.c

```diff
@@ -180,7 +191,105 @@ nsds5ReplicaBusyWaitTime - time to wait after getting a REPLICA BUSY from the co
 nsds5ReplicaSessionPauseTime - time to pause after sending updates to allow another supplier to send
 */
 
+/* It sets various fields related to client side (repl_agmt)
+ * of the session tracking
+ * - session_id_cnt (counting the the outbound connection)
+ * - session_tracking_supported (a flag stating if the server side support SID)
+ * - session_id_prefix (fixed part of the session identifier for this repl_agmt)
+ */
+static void
+agmt_init_session_id(Repl_Agmt *ra)
+{
+    char hash_out[HASH_LENGTH_MAX] = {0};
+    char *enc = NULL;
+    char *root = NULL; /* e.g. dc=example,dc=com */
+    char *host = NULL; /* e.g. localhost.domain */
+    char port[10];   /* e.g. 389 */
+    char sport[10];  /* e.g. 636 */
+    char *hash_in;
+
+    if (ra == NULL) {
+        goto fail;
+    }
+    ra->session_id_cnt = 1;
+    ra->session_tracking_supported = SID_SUPPORT_UNKNOWN;
+    root = slapi_ch_strdup(slapi_sdn_get_dn(ra->replarea));
+    if (root == NULL) {
+        root = slapi_ch_strdup("unknown suffix");
+    }
+    host = get_localhost_DNS();
+    if (host == NULL) {
+        host = slapi_ch_strdup("unknown host");
+    }
+    snprintf(port,  sizeof(port), "%d", config_get_port());
+    snprintf(sport, sizeof(sport),"%d", config_get_secureport());
+    hash_in = slapi_ch_calloc(1, strlen(root) + strlen(host) + strlen(port) + strlen(sport) + 1);
+    if (hash_in == NULL) {
+        goto fail;
+    }
+    sprintf(hash_in, "%s%s%s%s", root, host, port, sport);
+    PK11_HashBuf(SEC_OID_SHA1, (unsigned char *)hash_out, (unsigned char *)hash_in, strlen(hash_in));
+
+    if ((enc = slapi_ch_calloc(LDIF_BASE64_LEN(SHA1_LENGTH) + 1, sizeof(char))) == NULL) {
+        goto fail;
+    }
+    (void)PL_Base64Encode(hash_out, SHA1_LENGTH, enc);
+    goto done;
+
+fail:
+    enc = slapi_ch_strdup("dummyID");
+
+done:
+    if (hash_in) {
+        slapi_ch_free_string(&hash_in);
+    }
+    if (root) {
+        slapi_ch_free_string(&root);
+    }
+    if (host) {
+        slapi_ch_free_string(&host);
+    }
+    ra->session_id_prefix = enc;
+    strncpy(ra->session_id, ra->session_id_prefix, SESSION_ID_STR_SZ - 4);
+    strcpy(&ra->session_id[SESSION_ID_STR_SZ - 4], " ---");
+}
+
+void
+agmt_set_session_id(Repl_Agmt *ra)
+{
+    if (ra->session_id_cnt == 999) {
+        ra->session_id_cnt = 1;
+    } else {
+        ra->session_id_cnt++;
+    }
+    strncpy(ra->session_id, ra->session_id_prefix, SESSION_ID_STR_SZ - 4);
+    sprintf(&ra->session_id[SESSION_ID_STR_SZ - 4], " %3d", ra->session_id_cnt);
```

Comment:
Why all the padding?

    sid "OBWBnz8jniH   4"

Can't we just use:

    sprintf(&ra->session_id[SESSION_ID_STR_SZ - 4], " %d", ra->session_id_cnt);
   
    sid "OBWBnz8jniH 4"